nutrition_labels/grant_tagger.py

```diff
@@ -11,248 +39,358 @@
-        self.vectorizer_type = vectorizer_type
-        self.model_type = model_type
-        self.bert_type = bert_type
         self.relevant_sample_ratio = relevant_sample_ratio
+        self.split_seed = split_seed
+        self.vectorizer_type = vectorizer_type
+        self.classifier_type = classifier_type
+        self.prediction_cols = (*prediction_cols,)
+        self.label_name = label_name
 
-    def transform(self, data):
+    def transform(self, data, train_data_id="Internal ID"):
 
-        if 'Grant texts' not in data:
+        if "Grant texts" not in data:
             # If the training data hasn't come through Prodigy tagging then this won't exist
-            data['Grant texts'] = data[['Title', 'Grant Programme:Title', 'Description']].agg(
-                '. '.join, axis=1
-                )
-        self.X = data['Grant texts'].tolist()
-        y = data['Relevance code']
+            data["Grant texts"] = data[list(self.prediction_cols)].agg(
+                ". ".join, axis=1
+            )
+        self.X_ids = data[train_data_id].tolist()
+        self.X = data["Grant texts"].tolist()
+        self.y = data[self.label_name].tolist()
 
-        if self.vectorizer_type == 'count':
+        if self.vectorizer_type == "count":
             self.vectorizer = CountVectorizer(
-                analyzer='word',
-                token_pattern=r'(?u)\b\w+\b',
-                ngram_range=self.ngram_range
-                )
-        elif self.vectorizer_type == 'tfidf':
+                analyzer="word",
+                token_pattern=r"(?u)\b\w+\b",
+                ngram_range=(1, 2),
+            )
+        elif self.vectorizer_type == "tfidf":
             self.vectorizer = TfidfVectorizer(
-                analyzer='word',
-                token_pattern=r'(?u)\b\w+\b',
-                ngram_range=self.ngram_range
-                )
-        elif self.vectorizer_type == 'bert':
-            self.vectorizer = BertVectorizer(pretrained=self.bert_type)
+                analyzer="word",
+                token_pattern=r"(?u)\b\w+\b",
+                ngram_range=(1, 2),
+            )
+        elif "bert" in self.vectorizer_type:
+            self.vectorizer = BertVectorizer(pretrained=self.vectorizer_type)
         else:
-            print('Vectorizer type not recognised')
-        X_vect = self.vectorizer.fit_transform(self.X)
+            print("Vectorizer type not recognised")
+        self.X_vect = self.vectorizer.fit_transform(self.X)
 
-        if self.vectorizer_type == 'bert' and self.model_type == 'naive_bayes':
+        if "bert" in self.vectorizer_type and self.classifier_type == "naive_bayes":
             scaler = MinMaxScaler()
-            X_vect = scaler.fit_transform(X_vect)
+            self.X_vect = scaler.fit_transform(self.X_vect)
 
-        return X_vect, y
+        return self.X_vect, self.y
 
-    def split_data(self, X_vect, y, split_seed):
+    def split_data(self, X_vect, y):
```

Comment:
why split_data is part of the model? usually we split data before passing them to the model. also I notice that there seem to be some data that are deemed irrelevant, what are those? I wonder whether those along with splitting should be dealt in preprocessing before the model ever gets called